src/main/java/com/openshift/internal/restclient/okhttp/OpenShiftAuthenticator.java

```diff
@@ -0,0 +1,139 @@
+	private static final String ERROR = "error";
+	private static final String ERROR_DETAILS = "error_details";
+	
+	private Collection<IChallangeHandler> challangeHandlers = new ArrayList<>();
+	private OkHttpClient okClient;
+	private IClient client;
+	
+	@Override
+	public Request authenticate(Route route, Response response) throws IOException {
+		if(unauthorizedForCluster(response)){
+			String requestUrl = response.request().url().toString();
+			Request authRequest = new Request.Builder()
+					.addHeader("X-CSRF-Token", "1")
+					.url(route.address().url().toString() + "oauth/authorize?response_type=token&client_id=openshift-challenging-client")
+					.build();
+			try (
+				Response authResponse = tryAuth(authRequest)){
+				if(authResponse.isSuccessful()) {
+					String token = captureToken(authResponse);
+					return response.request().newBuilder()
+							.header(IHttpConstants.PROPERTY_AUTHORIZATION, String.format("%s %s", IHttpConstants.AUTHORIZATION_BEARER, token))
+							.build();
+				}
+			}
+			throw new UnauthorizedException(captureAuthDetails(requestUrl), ResponseCodeInterceptor.getStatus(response.body().string()));
+		}
+
+		return null;
+	}
+	
+	private boolean unauthorizedForCluster(Response response) {
+		String requestHost = response.request().url().host();
+		return response.code() == IHttpConstants.STATUS_UNAUTHORIZED && client.getBaseURL().getHost().equals(requestHost);
+	}
+	
+	private Response tryAuth(Request authRequest) throws IOException {
+		return okClient
+		.newBuilder()
+		.authenticator(new Authenticator() {
+			
+			@Override
+			public Request authenticate(Route route, Response response) throws IOException {
+				if(StringUtils.isNotBlank(response.request().header(AUTH_ATTEMPTS))) {
+					return null;
+				}
+				if(StringUtils.isNotBlank(response.header(IHttpConstants.PROPERTY_WWW_AUTHENTICATE))) {
+					for (IChallangeHandler challangeHandler : challangeHandlers) {
+						if(!challangeHandler.canHandle(response.headers())) {
+							Builder requestBuilder = response.request().newBuilder()
+									.header(AUTH_ATTEMPTS, "1");
+							return challangeHandler.handleChallange(requestBuilder).build();
+						}
+					}
+				}
+				return null;
+			}
+		})
+		.build()
+		.newCall(authRequest).execute();
+	}
+	
+	private IAuthorizationDetails captureAuthDetails(String url) {
+		IAuthorizationDetails details = null;
+		Map<String, String> pairs = URIUtils.splitFragment(url);
+		if (pairs.containsKey(ERROR)) {
+			details = new AuthorizationDetails(pairs.get(ERROR), pairs.get(ERROR_DETAILS));
+		}
+		return details;
+	}
+	
+	private String captureToken(Response response) {
+		String token = null;
+		Map<String, String> pairs = URIUtils.splitFragment(response.header(PROPERTY_LOCATION));
+		if (pairs.containsKey(ACCESS_TOKEN)) {
+			token = pairs.get(ACCESS_TOKEN);
+			IAuthorizationContext authContext = client.getAuthorizationContext();
+			if(authContext != null) {
```

Comment:
shouldnt we let the user know that we couldnt find the context? What's the situation in that case? Is it an error, does it mean than auth wont be possible or ...?